### PR TITLE
JSON resources, flat and nested

### PR DESCRIPTION
**Pull request by wpdebruin**
by changing resourceType setting from 'java' (default) to 'json' you will be able to use flat and nested json files.
The nested JSON is flattened on loading.

If we want to add more JSON formats, we could plug in more json import functions.

### CODE REVIEW

**Comment by wpdebruin** [comment]
have a few minor issues left before we are ready. I need some opinions:

We are setting the currentLocale in the storageService. But it is called DefaultLocale, e.g
`variables.storageService.set( "DefaultLocale", arguments.locale );`
But it is confusing. It is not the default Locale but the CURRENT LOCALE, so cbi18n can have a default and by `setFwLocale` you can change the current Locale to something different. It would make more sense to change it to something called currentLocale of storedLocale e.g
`variables.storageService.set( "currentLocale", arguments.locale );`

Now we can set a default Locale. If you don't do that you will end up with an empty string, which means no decent resource names and so on. Why are we not making 'en_US' the default in the settings?
We are already doing it in ONE place:
'	private function buildLocale( string thisLocale = "en_US" ) {'
so why not in the module settings. I wouldn't expect nl_NL to be the default, so as a non US person I would always set a default locale.

More important and already a modification on NEW code. Now we have inherited settings, so resources from main_en_US.properties merged with main_EN.properties merged with main.properties (or JSON). But I am ALSO loading the default locale properties which gets merged. I think now we shouldn't do that.  So if you would load main_es_SV it would now initially load main_en_US and merge it with main.properties, main_es.properties and man_es_SV properties if they are present. This has one flaw: all keys missing from main, main_es, main_es_SV will be automatically substituted by english keys from main_en_US.  I don't think we want that, we only want main, main_es and main_es_SV keys in our locale resource.

**Comment by lmajano** [comment]
I agree with chaning it to `currentLocale` from `defaultLocale` this was just chosen lots of years ago!

>Now we can set a default Locale. If you don't do that you will end up with an empty string, which means no decent resource names and so on. Why are we not making 'en_US' the default in the settings?

I agree.

>we only want main, main_es and main_es_SV keys in our locale resource.

I agree

File: models/ResourceService.cfc [hunk 1]
```diff
@@ -26,28 +30,34 @@ component singleton accessors="true" {
 	 * @controller.inject coldbox
 	 * @i18n cbi18n module
 	 * @i18n.inject i18n@cbi18n
+	 * @settings cbi18n settings
+	 * @settings.inject coldbox:moduleSettings:cbi18n
+	 * 
+	 * @throws cbi18n.InvalidConfiguration
 	 */
-	function init( required controller, required i18n ) {
+	function init( required controller, required i18n, required settings ) {
```
Comment (lmajano) [inline models/ResourceService.cfc:38]: I am wondering why did you move the settinghs to the constructor and not to the onDiComplete method?  Because now you have double injection going on, as properties and as constructors?

File: models/ResourceService.cfc [hunk 2]
```diff
@@ -342,29 +361,36 @@ component singleton accessors="true" {
...
+					arrayAppend( keys, rbKeys.nextElement() );
+				}
+			} finally {
+				fis.close();
 			}
-		} finally {
-			fis.close();
+			return keys;
+		} else {
+			var myResource= _flattenStruct( deserializeJSON( fileRead( "#arguments.rbFile#_#arguments.rbLocale#.json" ) ) );
```
Comment (lmajano) [inline models/ResourceService.cfc:389]: Why not call `_loadJsonSubBundle()` if you have that code there already?

File: models/ResourceService.cfc [hunk 3]
```diff
@@ -299,17 +313,22 @@ component singleton accessors="true" {
...
+			var fis = getResourceFileInputStream( "#arguments.rbFile#_#arguments.rbLocale#.properties" );
+			var rb  = createObject( "java", "java.util.PropertyResourceBundle" ).init( fis );
+			try {
+				// Retrieve string
+				var rbString = rb.handleGetObject( arguments.rbKey );
+			} finally {
+				fis.close();
+			}
+		} else {
+			var myJsonResource = _flattenStruct( deserializeJSON( fileRead( "#arguments.rbFile#_#arguments.rbLocale#.json" ) ) );
```
Comment (lmajano) [inline models/ResourceService.cfc:327]: Why not call `_loadJsonSubBundle()` if you have that code there already?

**Review by lmajano — CHANGES_REQUESTED** [review]

File: models/ResourceService.cfc (hunk 1 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:38]: I agree. It should be one or the other, not both. I was confused because I saw the constructor method in another module. But onDiComplete is cleaner, as lang as we are sure it will always be instantiated with Wirebox.
In qb default settings are injected in the object by creating some initArg stuff in a binder.map. But of course we can do the same with .property() or .setter(). I'll change it to onDiComplete again

**Comment by wpdebruin** [comment]
> I agree with chaning it to `currentLocale` from `defaultLocale` this was just chosen lots of years ago!
> 
Done
> > Now we can set a default Locale. If you don't do that you will end up with an empty string, which means no decent resource names and so on. Why are we not making 'en_US' the default in the settings?
> 
> I agree.
> 
Done
> > we only want main, main_es and main_es_SV keys in our locale resource.
> 
> I agree
Done

File: models/ResourceService.cfc (hunk 1 above)
Comment (wpdebruin) [reply models/ResourceService.cfc]: Ok, controller and i18n were already done in the constructor. I moved everything to property injections, created onDiComplete again.  Since we have a .settings property it didn't make sense to populate all these individual properties again from settings. So init and onDiComplete is a lot simpler now. (I think this was all left from the original instance.something style.
And modified the tests a little.

File: models/ResourceService.cfc (hunk 2 above)
Comment (wpdebruin) [reply models/ResourceService.cfc]: done

File: readme.md [hunk 4]
```diff
@@ -29,20 +29,21 @@ This module registers the following models in WireBox:
...
     defaultLocale = "en_US",
-    // The storage to use for user's locale: session, client, cookie, request
-    localeStorage = "cookie",
+    // The storage to use for user's locale: any cbstorages service. Please use full wirebox ID
+    localeStorage = "CookieStorage@cbstorages",
     // The value to show when a translation is not found
     unknownTranslation = "**NOT FOUND**",
     logUnknownTranslation = true | false,
-    // Extra resource bundles to load
+    // Extra resource bundles to load, specify path up to but not including _lang_COUNTRY.properties here
```
Comment (lmajano) [inline readme.md:46]: Can you add `resourceType` here as well please, it needs to be documented.

File: models/i18n.cfc [hunk 5]
```diff
@@ -165,86 +165,95 @@ component singleton accessors="true" {
...
 	 * returns list of locale names, UNICODE direction char (LRE/RLE) added as required
 	 */
-	string function getLocaleNames() {
-		var orgLocales   = getLocales();
+	string function getLocaleNames(){
 		var theseLocales = "";
 		var thisName     = "";
-		orgLocales.each( function( orgLocale ) {
-			if ( orgLocale.listLen( "_" ) == 2 ) {
+		for ( orgLocale in getLocales() ) {
```
Comment (lmajano) [inline models/i18n.cfc:190]: Why change this from the original `each`? This actually introduces a var scope issue with `orgLocale`.  Can we revert back.

File: models/ResourceService.cfc [hunk 6]
```diff
@@ -8,54 +8,51 @@
 component singleton accessors="true" {
 
 	// DI
-	property name="log"      inject="logbox:logger:{this}";
-	property name="i18n"     inject="i18n@cbi18n";
-	property name="settings" inject="coldbox:moduleSettings:cbi18n";
+	property name="log"        inject="logbox:logger:{this}";
+	property name="controller" inject="coldbox";
+	property name="i18n"       inject="provider:i18n@cbi18n";
```
Comment (lmajano) [inline models/ResourceService.cfc:13]: Good catch. Circular dep

File: models/ResourceService.cfc [hunk 7]
```diff
@@ -514,14 +579,17 @@ component singleton accessors="true" {
 	 * @return struct resourcebundle
 	 * @throws ResourceBundle.InvalidBundlePath
 	 */
-	private function _loadSubBundle( required string resourceBundleFullPath ) {
+	private function _loadJavaSubBundle( required string resourceBundleFullPath ){
 		var resourceBundle = {};
 		var thisKey        = "";
 		// create a file input stream with file location
 		var fis            = getResourceFileInputStream( resourceBundleFullPath );
```
Comment (lmajano) [inline models/ResourceService.cfc:586]: arguments.resourceBundleFullPath

**Review by lmajano — CHANGES_REQUESTED** [review]

**Comment by lmajano** [comment]
Hey @wpdebruin any update on this so we can finalize and release?

**Comment by wpdebruin** [comment]
Hey,

Sorry about that. There are still some circular serious dependency issues. I have to investigate this before we continue, but I had didn’t have much time this week. Working on a new project where I am using this code with JSON resources, but I see it is not error-free yet. Hope to fix it within a week.

Wil

Van: Luis Majano <notifications@github.com>
Beantwoorden - Aan: coldbox-modules/cbi18n <reply@reply.github.com>
Datum: donderdag 5 november 2020 om 00:43
Aan: coldbox-modules/cbi18n <cbi18n@noreply.github.com>
CC: "wil@site4u.nl" <wil@site4u.nl>, Mention <mention@noreply.github.com>
Onderwerp: Re: [coldbox-modules/cbi18n] JSON resources, flat and nested (#33)


Hey @wpdebruin<https://github.com/wpdebruin> any update on this so we can finalize and release?

File: models/i18n.cfc (hunk 5 above)
Comment (wpdebruin) [reply models/i18n.cfc:190]: orgLocales is actually getLocales which is a Java array, not a cf array. I like the cfarray.each syntax more, but I had many places where in  the code where I got undocumented responses from the original java array so I didn't want to change it anymore. Things like
```
	boolean function isValidLocale( required string thisLocale ){
		return (
			listFind(
				arrayToList( getLocales() ),
				arguments.thisLocale
			)
		) ? true : false;
	}
```
or this
```
		for ( var orgLocale in getLocales() ) {
			if ( listLen( orgLocale, "_" ) == 2 ) {
```
which had different results (or errors) if I changed getLocales result to a cf array. 
There are many more places where java arrays are returned, but I never heard anyone complain about that.
So I leave it for now, or we should further rewrite the module.

Comment (wpdebruin) [reply models/i18n.cfc:190]: I also rewrote getLocaleQuery() because it was broken (but not in tests)

File: models/ResourceService.cfc (hunk 7 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:586]: done

**Comment by lmajano** [comment]
@wpdebruin is this ready to be merged now?
I see allt ests pass now?
